helix-core/src/test/java/org/apache/helix/store/zk/TestZkHelixPropertyStore.java

```diff
@@ -88,6 +92,57 @@ public void afterClass() {
     deleteCluster(getShortClassName());
   }
 
+  @Test
+  public void testSessionExpirationWithSharedZkClient() throws Exception {
+    /*
+    This test is to make sure SharedZkClient would work with
+    HelixPropertyStore when session expiration happens. More specifically
+    HelixPropertyStore would register all the callbacks after session
+    expiration.
+     */
+    String subRoot = _root + "/" + "localCallback";
+    List<String> subscribedPaths = new ArrayList<>();
+    subscribedPaths.add(subRoot);
+
+    ZkSerializer serializer = new ZNRecordSerializer();
+    ZkHelixPropertyStore<ZNRecord> store =
+        new ZkHelixPropertyStore<>(ZK_ADDR, serializer, subRoot, subscribedPaths);
+
+    TestListener listener = new TestListener();
+    store.subscribe("/", listener);
+
+    // test dataCreate callbacks
+    listener.reset();
+    setNodes(store, 'a', true);
+
+    // kill the session to make sure shared zkClient re-installs watcher
+    // Note, current shared ZkClient does not issue new session. Thus, we need
+    // kill _zkConnectionManager's session. Otherwise, ZkTestHelper.expireSesson
+    // would not work.
+    SharedZkClient sharedClient = (SharedZkClient) store.getTestZkClient();
+    HelixZkClient  testClient =  sharedClient.getConnectionManager();
+    ZkTestHelper.expireSession(testClient);
+
+
+    // kill the session one more time to cover Shared ZkClient resetting flag
+    // indicating first time synconnect happened.
+    ZkTestHelper.expireSession(testClient);
+
+    listener.reset();
+    int expectDeleteNodes = 1 + firstLevelNr + firstLevelNr * secondLevelNr;
+    store.remove("/", 0);
+    // wait until all callbacks have been received
+    for (int i = 0; i < 10; i++) {
+      if (listener._deleteKeys.size() == expectDeleteNodes)
+        break;
+      Thread.sleep(500);
```

Comment:
`sleep()` doesn't look solid and makes the test unstable. I suggest using `TestHelper.varify()` to replace sleep().

Reply:
+1. This is not a show of good craftsmanship..

Reply:
This is copied from the other test. I can change this a little bit.

Reply:
New approach use TestHelper.verifyWithTimeout().